torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  if ((shape.size() - start) != ndim) {
+    return false;
+  }
+
+  npy_intp* array_shape = PyArray_DIMS(array);
+  for (int64_t i = 0; i < ndim; ++i) {
+    if (array_shape[i] != shape[i+start]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+
+inline int _validate_ndarray(PyObject* obj, std::vector<int64_t> & shape, std::vector<int64_t> & ndarray_shape, int64_t depth) {
+  /*
+  This returns -1 if obj is a scalar array, 0 if obj is not array else if 1 given no exception is raised.
+  A ValueError will be raised if the depth of obj within the input iterable, as given to _extract_ndarrays,
+  is not the same as any other array within the input iterable OR if its shape does not match any other
+  array's shape in the iterable.
+  */
+
+  PyArrayObject* array;
+  if (PyArray_Check(obj)) {
+    array = (PyArrayObject*) obj;
+  } else {
+    return 0;
+  } 
+
+  if (PyArray_NDIM(array) == 0) {
+    return -1;
+  }
+
+  if (ndarray_shape.size() == 0) {
+    /*
+    If ndarray_shape is an empty vector, then ndarray_shape[0] is set to the depth of the array in the iterable and
+    ndarray_shape[1:] is set to the shape of array.
+    */
+    ndarray_shape.emplace_back(depth);
+    int64_t ndim = PyArray_NDIM(array);
+    npy_intp* array_shape = PyArray_DIMS(array);
+    ndarray_shape.insert(ndarray_shape.end(), array_shape, array_shape + ndim);
+  }
+  else if ((ndarray_shape[0] != depth) || !_shape_match_ndarray(ndarray_shape, array, 1)) {
+    /*
+    If the depth of 'array' in the iterable is not the same as any other array's depth OR if the shape of 'array' does not
+    match with any other array's shape in the iterable, we raise a ValueError.
+    */
+    auto true_shape_str = get_buffer_str<npy_intp> (static_cast<void*>(PyArray_DIMS(array)), 0, PyArray_NDIM(array));
+    auto expected_shape_str = get_vector_str(ndarray_shape, 1);
+    std::ostringstream err;
+    err << "expected numpy array of shape " << expected_shape_str <<" at dim " << ndarray_shape[0] << " (got " << true_shape_str << " at dim " << depth << ")";
+    throw ValueError(err.str());  
+  }
+
+  return 1;
+}
+
+bool _extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, std::vector<int64_t> & ndarray_shape, int64_t depth, std::vector<PyArrayObject*> & array_ptr_storage) {
+  /*
+  This is a helper function for extracting all the numpy arrays embedded within PySequences (recursively) representd by obj. This is done recursively in a 
+  depth first search style. It puts the numpy array pointers in the array_ptr_storage for later use. In case there is a non numpy array type within obj, 
+  this function immediately returns false. If only all the items inside obj are just non-scalar numpy arrays, it returns true. It throws a ValueError in case
+  of a dimension and/or depth mismatch (See _validate_ndarray for more info). This function is used in extract_ndarrays() function below.
+  */
+
+  int validation_flag = _validate_ndarray(obj, shape, ndarray_shape, depth);
```

Comment:
One idea to improve readability is to split this into acquiring the first array and its parameters (depth and shape), then have the next function discover all the arrays (and validate them when they're discovered), then have a third function operate on the discovered arrays.

Reply:
See https://github.com/pytorch/pytorch/pull/51731#discussion_r602715434